### ISSUE TITLE
`clippy::option_if_let_else` suggests invalid code fix

### ISSUE THREAD

**Issue by u9g** [rust-lang/rust-clippy#11059]
### Summary

The code produced by clippy is in theory correct in my eyes, but in reality the type checker doesn't understand it.

### Reproducer

I tried this code:

```rust
if let Some(id) = &v.as_class().unwrap().id {
    Box::new(std::iter::once(Vertex::Span(id.span)))
} else {
    Box::new(std::iter::empty())
}
```

I expected to see this happen: to get working code after a quick fix.
Instead, this happened: The quick fix produces the following non-working code:
```rust
v.as_class().unwrap().id.as_ref().map_or_else(
    || Box::new(std::iter::empty()),
    |id| Box::new(std::iter::once(Vertex::Span(id.span))),
)
```
Which is invalid because:
```
error[E0308]: mismatched types
   --> crates\oxc_linter\src\lint_adapter2.rs:217:35
    |
217 |                     |id| Box::new(std::iter::once(Vertex::Span(id.span))),
    |                          -------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Empty<_>`, found `Once<Vertex<'_>>`
    |                          |
    |                          arguments to this function are incorrect
    |
    = note: expected struct `std::iter::Empty<_>`
               found struct `std::iter::Once<lint_adapter2::Vertex<'_>>`
note: associated function defined here
   --> C:\Users\jason\.rustup\toolchains\nightly-2023-06-02-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:215:12
    |
215 |     pub fn new(x: T) -> Self {
    |            ^^^

```

### Version

```text
rustc 1.72.0-nightly (d59363ad0 2023-06-01)
binary: rustc
commit-hash: d59363ad0b6391b7fc5bbb02c9ccf9300eef3753
commit-date: 2023-06-01
host: x86_64-pc-windows-msvc
release: 1.72.0-nightly
LLVM version: 16.0.4
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by zaneduffield** [comment]
Hi @u9g, I think your original code doesn't type-check either because the `if` and `else` branches have incompatible types
```
  | |_____- `if` and `else` have incompatible types
  |
  = note: expected struct `std::boxed::Box<std::iter::Once<&_>>`
             found struct `std::boxed::Box<std::iter::Empty<_>>`
```

**Comment by y21** [comment]
@zaneduffield it depends in which position this piece of code is. This for example does pass typeck, because the two types are coerced into trait objects because of the return type:
```rs
use std::fmt::Debug;

fn foo(o: Option<i32>) -> Box<dyn Debug> {
  if let Some(o) = o {
    Box::new(o)
  } else {
    Box::new("foo")
  }
}
```
But this coercion doesn't happen implicitly in `Option::map_or_else` and would need an explicit `as`-cast to `Box<dyn Debug>`
